loader_hub/readwise/base.py

```diff
@@ -0,0 +1,46 @@
+"""Simple Reader that loads highlights from readwise.io"""
+import requests
+import json
+from typing import List
+
+from llama_index.readers.base import BaseReader
+from llama_index.readers.schema.base import Document
+
+def _get_readwise_data(api_key, updated_after=None):
+  """
+  Uses Readwise's export API to export all highlights, optionally after a specified date.
+
+  See https://readwise.io/api_deets for details.
+
+  Args:
+      updated_after (datetime.datetime): The datetime to load highlights after. Useful for updating indexes over time.
+  """
+  next_page = None
+  while True:
+    response = requests.get(
+      url="https://readwise.io/api/v2/export/",
+      params={"pageCursor": next_page, "updatedAfter": updated_after},
+      headers={"Authorization": f"Token {api_key}"})
+    response.raise_for_status()
+    yield from response.json()["results"]
+    next_page = response.json().get("nextPageCursor")
+    if not next_page: break
+
+class ReadwiseReader(BaseReader):
+    """
+    Reader for Readwise highlights.
+    """
+    def __init__(self, api_key):
+        self._api_key = api_key
+
+    def load_data(
+        self,
+        updated_after = None,
+    ) -> List[Document]:
+        """
+        Load your Readwise.io highlights.
+
+        Args:
+            updated_after (datetime.datetime): The datetime to load highlights after. Useful for updating indexes over time.
+        """
+        return [Document(json.dumps(d)) for d in _get_readwise_data(api_key=self._api_key, updated_after=updated_after)]
```

Comment:
I'm getting the following issue when I test this.

Here is the test code:

```python
import os
from llama_index import GPTSimpleVectorIndex, download_loader

URL = "https://raw.githubusercontent.com/alexbowe/llama-hub-readwise/main/loader_hub"
ReadwiseReader = download_loader("ReadwiseReader", loader_hub_url=URL)
token = os.getenv("READWISE_API_KEY")
loader = ReadwiseReader(api_key=token)
documents = loader.load_data()
print(type(documents[0]))
index = GPTSimpleVectorIndex(documents)
index.query("What was the paper 'Attention is all you need' about?")
```

And the output:

```
Traceback (most recent call last):
  File "main.py", line 10, in <module>
    index = GPTSimpleVectorIndex(documents)
  File "/home/runner/LawngreenGargantuanProcedurallanguage/venv/lib/python3.10/site-packages/llama_index/indices/vector_store/vector_indices.py", line 83, in __init__
    super().__init__(
  File "/home/runner/LawngreenGargantuanProcedurallanguage/venv/lib/python3.10/site-packages/llama_index/indices/vector_store/base.py", line 62, in __init__
    super().__init__(
  File "/home/runner/LawngreenGargantuanProcedurallanguage/venv/lib/python3.10/site-packages/llama_index/indices/base.py", line 104, in __init__
    documents = self._process_documents(
  File "/home/runner/LawngreenGargantuanProcedurallanguage/venv/lib/python3.10/site-packages/llama_index/indices/base.py", line 181, in _process_documents
    raise ValueError(f"Invalid document type: {type(doc)}.")
ValueError: Invalid document type: <class 'dict'>.
```

I'm not sure why yet. Will look into it more.

Reply:
@jerryjliu any idea why I might be getting this error even though I'm returning a list of `Document` objects?

Reply:
not sure, could you try calling get_text() on each Document object to make sure the text is set?

Reply:
@jerryjliu I think I set the `loader_hub_url` incorrectly.

In your code you have:

https://github.com/jerryjliu/gpt_index/blob/main/gpt_index/readers/download.py#L17-L19

Which would give `"https://raw.githubusercontent.com/emptycrown/loader-hub/main/loader_hub"`. So I was trying to form a similar link (`"https://raw.githubusercontent.com/alexbowe/llama-hub-readwise/main/loader_hub"`). Both of these end up with 404 not found errors when visited directly.

How can I find the correct link I need to test this?

Reply:
Tested it without using the hub loader URL. Would be great to update the Readme with how to do this at some point though :)